### ISSUE TITLE
[core] Remove the `Option<State>` around all state machines

### ISSUE THREAD

**Issue by piegamesde** [piegamesde/magic-wormhole.rs#4]
The state machines are mostly structs with an `state: Option<State>`, where `State` is an enum. The `None` state is never used however. It probably only exists to work around some type system restriction. I think that this can be removed.

The `Option` was introduced in a series of commits titled ["simplify state machines"](https://github.com/piegamesde/magic-wormhole.rs/commit/d29c6a0a9f1e818b84f2bdb60b56d13b2d019d4c) a long time ago. Before changing this, we should look at that change, what exactly was changed and why.

**Comment by piegamesde** [comment]
The problem is that the state machine requires temporarily moving a values out of a borrowed reference. In Rust, this is not allowed. Possible solutions to this are:

- Make sure all state is `Copy` (unpractical for a lot of our use cases)
- Liberally `Clone` the value (same)
- Use the `take_mut` crate (it does some unsafe stuff internally and provides a safe API around it)
- Use an `Option` to emulate `take_mut` functionality with slightly weaker type system guarantees (this is what we do now)
- Emulate what the `take_mut` crate does now but using an `Option`. This doesn't require us to add yet another dependency, but I can't tell how well this approach is going to work before having trying it out.

**Comment by piegamesde** [comment]
More or less obsoleted through #9. The code now uses a mixture of multiple things:

- On a plain enum state `Copy` is a valid option
- Take and return the parameter by move (`self` -> `Self`). This works really great on a lot of places.
- I also tried out a special wrapper around `Option`. It's only a few dozen lines of boilerplate code. It did work out rather well, but in the end I decided that it isn't worth it and that using an `Option` in the remaining cases is fine.
